### PR TITLE
GLTF: Add a comment for skinned mesh tree placement

### PR DESCRIPTION
**Pull request by aaronfranke**
This PR adds a comment explaining why we generate an empty node here. The reason is not obvious from just reading the code, so we need to have a comment explaining why we do this.

### CODE REVIEW

**Comment by fire** [comment]
Since this is sensitive code, I think adding the comment is enough.

I am not sure the code compression is worth it.

**Comment by fire** [comment]
We discussed there's a tradeoff between small bug fixes and large overhauls. Since this code is load bearing, smaller changes aren't worth doing the validating for correctness.

@aaronfranke wanted to make a new proposal.

> The current discussion is about enhancing the Godot game engine's handling of GLTF nodes. Presently, Godot can only generate one node from a single GLTF node JSON, which limits the creation of complex structures without code duplication.
>
> The proposed solution is to modify the code to allow multiple Godot nodes from a single GLTF node JSON. This would enable creating intricate structures like a physics body with various components without repeating code.
> 
> However, there are debates on the implementation. One idea is to run all extensions first, and the first extension that provides a node directly maps to the GLTF node. The rest become children. If no extension provides a node, another element (camera, light, etc.) may be generated. A mesh is created last as it might want to be a leaf (like with skinned meshes).
> 
> There's also concern about backward compatibility. Some believe the old method should not be preserved, while others suggest an import option for backward compatibility. The consensus is to treat this as a significant version change that upgrades with a checkbox. While the need for change is agreed upon, the specifics are still under discussion.
> 
> **Summarized by ai**

**Review by RevoluPowered — APPROVED** [review]

File: modules/gltf/gltf_document.cpp [hunk 1]
```diff
@@ -5804,10 +5804,15 @@ void GLTFDocument::_generate_scene_node(Ref<GLTFState> p_state, const GLTFNodeIn
 	// If none of our GLTFDocumentExtension classes generated us a node, we generate one.
 	if (!current_node) {
 		if (gltf_node->skin >= 0 && gltf_node->mesh >= 0 && !gltf_node->children.is_empty()) {
+			// GLTF specifies that skinned meshes should ignore their node
+			// transforms, only being controlled by the skeleton, so Godot
+			// will a reparent a skinned mesh to its skeleton. However, we
+			// still need to ensure any child nodes keep their place in the
+			// tree, so if there are any child nodes, the skinned mesh must
+			// not be the base node, so generate an empty spatial base.
```
Comment (AThousandShips) [inline modules/gltf/gltf_document.cpp:5812]:
```suggestion
			// GLTF specifies that skinned meshes should ignore their node
			// transforms, only being controlled by the skeleton, so Godot
			// will reparent a skinned mesh to its skeleton. However, we
			// still need to ensure any child nodes keep their place in the
			// tree, so if there are any child nodes, the skinned mesh must
			// not be the base node, so generate an empty spatial base.
```
This should be regrouped IMO, it's very hard to read when each sentence is broken up between lines, especially as the condition above is wider so the width isn't really significant

**Review by fire — CHANGES_REQUESTED** [review]
We discussed there's a tradeoff between small bug fixes and large overhauls. Since this code is load bearing, smaller changes aren't worth doing the validating for correctness.

@aaronfranke has a proposal to modify the code to allow multiple Godot nodes from a single GLTF node JSON.

**Review by fire — APPROVED** [review]
@lyuma I think this is good now. Can you check?

File: modules/gltf/gltf_document.cpp [hunk 2]
```diff
@@ -5802,10 +5802,15 @@ void GLTFDocument::_generate_scene_node(Ref<GLTFState> p_state, const GLTFNodeIn
 	// If none of our GLTFDocumentExtension classes generated us a node, we generate one.
 	if (!current_node) {
 		if (gltf_node->skin >= 0 && gltf_node->mesh >= 0 && !gltf_node->children.is_empty()) {
+			// GLTF specifies that skinned meshes should ignore their node
+			// transforms, only being controlled by the skeleton, so Godot
+			// will reparent a skinned mesh to its skeleton. However, we
+			// still need to ensure any child nodes keep their place in the
+			// tree, so if there are any child nodes, the skinned mesh must
+			// not be the base node, so generate an empty spatial base.
```
Comment (AThousandShips) [inline modules/gltf/gltf_document.cpp:5810]:
```suggestion
			// GLTF specifies that skinned meshes should ignore their node transforms,
			// only being controlled by the skeleton, so Godot will reparent a skinned mesh to its skeleton.
			// However, we still need to ensure any child nodes keep their place in the tree,
			// so if there are any child nodes, the skinned mesh must not be the base node,
			// so generate an empty spatial base.
```
Or similar, the current layout is very hard to read

**Review by AThousandShips — APPROVED** [review]

**Review by lyuma — APPROVED** [review]
Looks great. This was indeed one of the trickier edge cases and you'd otherwise have to git blame to find the PR and the linked issue with the particular testcase that led this to be added.

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**